engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/JsForeignNode.java

```diff
@@ -0,0 +1,99 @@
+package org.enso.interpreter.epb;
+
+import com.oracle.truffle.api.dsl.NodeField;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropException;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import java.util.Arrays;
+
+/** A node responsible for performing foreign JS calls. */
+@NodeField(name = "foreignFunction", type = Object.class)
+@NodeField(name = "arity", type = int.class)
+abstract class JsForeignNode extends ForeignFunctionCallNode {
+
+  private @Child CoercePrimitiveNode coercePrimitiveNode = CoercePrimitiveNode.build();
+
+  abstract Object getForeignFunction();
+
+  abstract int getArity();
+
+  /**
+   * Creates a new instance of this node.
+   *
+   * @param argumentsCount the number of arguments the function expects (including {@code this})
+   * @param jsFunction the parsed JS object (required to be {@link
+   *     InteropLibrary#isExecutable(Object)})
+   * @return a node able to call the JS function with given arguments
+   */
+  static JsForeignNode build(Object jsFunction, int argumentsCount) {
+    return JsForeignNodeGen.create(jsFunction, argumentsCount);
+  }
+
+  @Specialization
+  Object doExecute(Object[] arguments, @CachedLibrary("foreignFunction") InteropLibrary iop)
+      throws InteropException {
+    var args = new ArgumentsArray(arguments);
+    var self = arguments[0];
+    var raw = iop.invokeMember(getForeignFunction(), "apply", self, args);
```

Comment:
I'd like to get rid of (most of) `JsForeignNode`, but I cannot. This `apply` trick to set `this` is one (small) reason.

The bigger problem is that Graal.js doesn't like `return` inside of _parse with arguments_ that is used by `parseGeneric` - that's in my opinion fixable on Graal.js side.